src/groupeddataframe/groupeddataframe.jl

```diff
@@ -34,15 +50,19 @@ end
 function Base.getproperty(gd::GroupedDataFrame, f::Symbol)
     if f in (:idx, :starts, :ends)
         # Group indices are computed lazily the first time they are accessed
+        Threads.lock(gd.lazy_lock)
```

Comment:
Thinking about this again, shouldn't this line be moved inside `if getfield(gd, f) === nothing`? Ideally threads wouldn't have to lock each other when accessing indices *once they have been computed*. To avoid threads from computing the fields multiple times (which would be wasteful, though probably not problematic), we could check `getfield(gd, f) === nothing` again after calling `lock`.

Reply:
Yes - I was also thinking about it. The solution is exactly as you say:

> we could check `getfield(gd, f) === nothing` again after calling lock.

I just wanted to have a simpler implementation as `lock`/`unlock` is cheap:
```
julia> function f(l)
       lock(l)
       unlock(l)
       end
f (generic function with 1 method)

julia> l = Threads.ReentrantLock()
ReentrantLock(nothing, Base.GenericCondition{Base.Threads.SpinLock}(Base.InvasiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), 0)

julia> f(l)

julia> @benchmark f($l)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     26.066 ns (0.00% GC)
  median time:      26.087 ns (0.00% GC)
  mean time:        27.893 ns (0.00% GC)
  maximum time:     77.285 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     993

```
and with threads often things are more tricky than they seem.

I will open a PR with the change you propose and we can discuss there.